File: sdks/java/src/main/java/se/kommunsign/sdk/KommunSignClient.java
package se.kommunsign.sdk;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.Objects;
import java.util.function.Supplier;

/** OpenAPI version 2026-08-02.2. */
public final class KommunSignClient {
  public static final String OPENAPI_VERSION = "2026-08-02.2";
  private final HttpClient http;
  private final URI baseUri;
  private final Supplier<String> accessToken;
  public KommunSignClient(HttpClient http, URI baseUri, Supplier<String> accessToken) {
    this.http = Objects.requireNonNull(http); this.baseUri = Objects.requireNonNull(baseUri); this.accessToken = Objects.requireNonNull(accessToken);
  }
  public HttpResponse<String> listSignatureCases() throws Exception { return send("GET", "signature-cases", null, null, null); }
  public HttpResponse<String> getSignatureCase(String id) throws Exception { return send("GET", "signature-cases/" + id, null, null, null); }
  public HttpResponse<String> createSignatureCase(String json, String key) throws Exception { return send("POST", "signature-cases", json, key, null); }
  public HttpResponse<String> sendSignatureCase(String id, String key, Long version) throws Exception { return send("POST", "signature-cases/" + id + "/send", null, key, version); }
  private HttpResponse<String> send(String method, String path, String json, String key, Long version) throws Exception {
    var builder = HttpRequest.newBuilder(baseUri.resolve(path)).header("Authorization", "Bearer " + accessToken.get());
    if (key != null) builder.header("Idempotency-Key", key);
    if (version != null) builder.header("If-Match", version.toString());
    if (json != null) builder.header("Content-Type", "application/json").method(method, HttpRequest.BodyPublishers.ofString(json));
    else builder.method(method, HttpRequest.BodyPublishers.noBody());
    var response = http.send(builder.build(), HttpResponse.BodyHandlers.ofString());
    if (response.statusCode() < 200 || response.statusCode() >= 300) throw new IllegalStateException("KommunSign API returned " + response.statusCode());
    return response;
  }
}

File: sdks/csharp/src/KommunSignClient.cs
// OpenAPI version: 2026-08-03.2
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;

namespace KommunSign.Sdk;
public sealed class KommunSignClient(HttpClient http, Func<CancellationToken, Task<string>> accessToken)
{
    public const string OpenApiVersion = "2026-08-03.2";
    public Task<JsonDocument> ListSignatureCasesAsync(CancellationToken ct = default) => SendAsync(HttpMethod.Get, "signature-cases", null, null, null, ct);
    public Task<JsonDocument> GetSignatureCaseAsync(Guid id, CancellationToken ct = default) => SendAsync(HttpMethod.Get, $"signature-cases/{id}", null, null, null, ct);
    public Task<JsonDocument> CreateSignatureCaseAsync(object body, string idempotencyKey, CancellationToken ct = default) => SendAsync(HttpMethod.Post, "signature-cases", body, idempotencyKey, null, ct);
    public Task<JsonDocument> AddDocumentAsync(Guid id, object body, string key, CancellationToken ct = default) => SendAsync(HttpMethod.Post, $"signature-cases/{id}/documents", body, key, null, ct);
    public Task<JsonDocument> AddSignerAsync(Guid id, object body, string key, CancellationToken ct = default) => SendAsync(HttpMethod.Post, $"signature-cases/{id}/signers", body, key, null, ct);
    public Task<JsonDocument> UpdateSignerAsync(Guid id, Guid signerId, object body, string key, long? version = null, CancellationToken ct = default) => SendAsync(HttpMethod.Patch, $"signature-cases/{id}/signers/{signerId}", body, key, version, ct);
    public Task<JsonDocument> CreateUploadAsync(object body, string key, CancellationToken ct = default) => SendAsync(HttpMethod.Post, "uploads", body, key, null, ct);
    public Task<JsonDocument> CompleteUploadAsync(Guid uploadId, string sha256, string key, CancellationToken ct = default) => SendAsync(HttpMethod.Post, $"uploads/{uploadId}/complete", new { sha256 }, key, null, ct);
    public Task<JsonDocument> SendCaseAsync(Guid id, string key, long? version = null, CancellationToken ct = default) => SendAsync(HttpMethod.Post, $"signature-cases/{id}/send", null, key, version, ct);
    public Task<JsonDocument> CancelCaseAsync(Guid id, string key, long? version = null, CancellationToken ct = default) => SendAsync(HttpMethod.Post, $"signature-cases/{id}/cancel", null, key, version, ct);
    private async Task<JsonDocument> SendAsync(HttpMethod method, string path, object? body, string? key, long? version, CancellationToken ct)
    {
        using var request = new HttpRequestMessage(method, path);
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", await accessToken(ct));
        if (key is not null) request.Headers.Add("Idempotency-Key", key);
        if (version is not null) request.Headers.TryAddWithoutValidation("If-Match", version.Value.ToString());
        if (body is not null) request.Content = new StringContent(JsonSerializer.Serialize(body), Encoding.UTF8, "application/json");
        using var response = await http.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, ct);
        var bytes = await response.Content.ReadAsByteArrayAsync(ct);
        if (!response.IsSuccessStatusCode) throw new HttpRequestException($"KommunSign API returned {(int)response.StatusCode}");
        return JsonDocument.Parse(bytes);
    }
}
